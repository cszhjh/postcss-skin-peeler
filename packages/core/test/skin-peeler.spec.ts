import { dirname, join, relative } from 'node:path'
import mockFs from 'mock-fs'
import postcss from 'postcss'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import plugin from '../src/main'
import type { PluginOptions } from '../src/types'

const FROM = join(__dirname, './style/index.css')
const IMG_SRC = join(__dirname, './images')
const SKIN_IMG = join(__dirname, './skin')
const ORIGIN_IMG_ONE = join(IMG_SRC, './one.png')
const ORIGIN_IMG_TWO = join(IMG_SRC, './two.png')
const SKIN_IMG_ONE = join(SKIN_IMG, './one.png')
const SKIN_IMG_TWO = join(SKIN_IMG, './two.png')
const ORIGIN_IMG_ONE_RELATIVE = relative(dirname(FROM), ORIGIN_IMG_ONE)
const ORIGIN_IMG_TWO_RELATIVE = relative(dirname(FROM), ORIGIN_IMG_TWO)

const PLUGIN_OPTIONS: PluginOptions = {
  imgSrc: IMG_SRC,
  skinSrc: SKIN_IMG,
  prefixSelector: '.skin',
}

async function run(css: string) {
  const { css: output } = await postcss([plugin(PLUGIN_OPTIONS)]).process(css, { from: FROM })
  return output
}

beforeAll(() => {
  mockFs({
    [ORIGIN_IMG_ONE]: Buffer.from([8, 6, 7, 5, 3, 0, 9]),
    [ORIGIN_IMG_TWO]: Buffer.from([9, 6, 7, 5, 3, 0, 9]),
    [SKIN_IMG_ONE]: Buffer.from([9, 7, 8, 6, 4, 1, 10]),
    [SKIN_IMG_TWO]: Buffer.from([10, 7, 8, 6, 4, 1, 10]),
  })
})

afterAll(() => {
  mockFs.restore()
})

describe('test url quotation mark', () => {
  it('test url without quotes', async () => {
    const input = `
      .non-quotation {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test url with single quotes', async () => {
    const input = `
      .single-quote {
        background: url('${ORIGIN_IMG_ONE_RELATIVE}');
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test url with double quotes', async () => {
    const input = `
      .double-quote {
        background: url("${ORIGIN_IMG_ONE_RELATIVE}");
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})

describe('test url with protocol', () => {
  it('test url with http', async () => {
    const input = `
      .http {
        background: url(http://varletjs.org/varlet_icon.png);
      }
    `
    const output = await run(input)
    expect(output).toBe(input)
  })

  it('test url with https', async () => {
    const input = `
      .http {
        background: url(https://varletjs.org/varlet_icon.png);
      }
    `
    const output = await run(input)
    expect(output).toBe(input)
  })
})

describe('test background stack', () => {
  it('test background is stacked in the single rule', async () => {
    const input = `
      .rule-single-stacked {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
        background-image: url(${ORIGIN_IMG_TWO_RELATIVE})
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test background is stacked in the multiple rule', async () => {
    const input = `
      .rule-multiple-stacked {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }

      .rule-multiple-stacked {
        background: url(${ORIGIN_IMG_TWO_RELATIVE});
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})

describe('test html and body combination', () => {
  it('test html as prefix selector', async () => {
    const input = `
      html .html-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test body as prefix selector', async () => {
    const input = `
      body .body-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test html body as prefix selector', async () => {
    const input = `
      html body .html-body-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test html follow selector as prefix selector', async () => {
    const input = `
      html.follow .html-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test body follow selector as prefix selector', async () => {
    const input = `
      body.follow .body-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})

describe('test rule in @media', () => {
  const input = `
    @media screen and (min-width: 1024px) {
      .media {
        background: url(${ORIGIN_IMG_ONE_RELATIVE});
      }
    }
  `
  it('test rule in @media', async () => {
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})
