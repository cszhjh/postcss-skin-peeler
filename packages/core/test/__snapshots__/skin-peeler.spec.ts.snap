// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test background stack > test background is stacked in the multiple rule 1`] = `
"
      .rule-multiple-stacked {
        background: url(../images/one.png);
      }
      .skin .rule-multiple-stacked {background-image: url("../skin/two.png");
      }

      .rule-multiple-stacked {
        background: url(../images/two.png);
      }
    "
`;

exports[`test background stack > test background is stacked in the single rule 1`] = `
"
      .rule-single-stacked {
        background: url(../images/one.png);
        background-image: url(../images/two.png)
      }
      .skin .rule-single-stacked {background-image: url("../skin/two.png")
      }
    "
`;

exports[`test html and body combination > test body as prefix selector 1`] = `
"
      body .body-prefix {
        background: url(../images/one.png);
      }
      body .skin .body-prefix {background-image: url("../skin/one.png");
      }
    "
`;

exports[`test html and body combination > test body follow selector as prefix selector 1`] = `
"
      body.follow .body-prefix {
        background: url(../images/one.png);
      }
      body.follow .skin .body-prefix {background-image: url("../skin/one.png");
      }
    "
`;

exports[`test html and body combination > test html as prefix selector 1`] = `
"
      html .html-prefix {
        background: url(../images/one.png);
      }
      html .skin .html-prefix {background-image: url("../skin/one.png");
      }
    "
`;

exports[`test html and body combination > test html body as prefix selector 1`] = `
"
      html body .html-body-prefix {
        background: url(../images/one.png);
      }
      html body .skin .html-body-prefix {background-image: url("../skin/one.png");
      }
    "
`;

exports[`test html and body combination > test html follow selector as prefix selector 1`] = `
"
      html.follow .html-prefix {
        background: url(../images/one.png);
      }
      html.follow .skin .html-prefix {background-image: url("../skin/one.png");
      }
    "
`;

exports[`test rule in @media > test rule in @media 1`] = `
"
    @media screen and (min-width: 1024px) {
      .media {
        background: url(../images/one.png);
      }
      .skin .media {background-image: url("../skin/one.png");
      }
    }
  "
`;

exports[`test url quotation mark > test url with double quotes 1`] = `
"
      .double-quote {
        background: url("../images/one.png");
      }
      .skin .double-quote {background-image: url("../skin/one.png");
      }
    "
`;

exports[`test url quotation mark > test url with single quotes 1`] = `
"
      .single-quote {
        background: url('../images/one.png');
      }
      .skin .single-quote {background-image: url("../skin/one.png");
      }
    "
`;

exports[`test url quotation mark > test url without quotes 1`] = `
"
      .non-quotation {
        background: url(../images/one.png);
      }
      .skin .non-quotation {background-image: url("../skin/one.png");
      }
    "
`;
